internal/rotate_est.go

```diff
@@ -0,0 +1,178 @@
+package internal
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"go.mozilla.org/pkcs7"
+)
+
+var (
+	oidKeyUsage         = asn1.ObjectIdentifier([]int{2, 5, 29, 15})
+	oidExtendedKeyUsage = asn1.ObjectIdentifier([]int{2, 5, 29, 37})
+
+	asn1DigitalSignature = []byte{3, 2, 7, 128}
+	asn1TlsWebClientAuth = []byte{48, 10, 6, 8, 43, 6, 1, 5, 5, 7, 3, 2}
+)
+
+type estStep struct{}
+
+func (s estStep) Name() string {
+	return "Generate new certificate"
+}
+
+func (s estStep) Completed(state CertRotationState) bool {
+	return len(state.NewKey) > 0 && len(state.NewCert) > 0
+}
+
+func (s estStep) Execute(handler *CertRotationHandler) error {
+	// Find the current certificate so we can build the proper EST request
+	tlsCert := handler.client.Transport.(*http.Transport).TLSClientConfig.Certificates[0]
+	cert, err := x509.ParseCertificate(tlsCert.Certificate[0])
+	if err != nil {
+		return nil
+	}
+
+	var signer crypto.Signer
+	var newKey string
+
+	// Generate a new private key
+	if handler.usePkcs11() {
+		newKey = "07"
```

Comment:
What does `"07"` mean?

Reply:
the `07` and `09` are new slot ids in the HSM where we can put objects. I need to clean this up though because it will only work correctly for one cert rotation. The next rotation, we'd accidentally overwrite the objects in use. I need to make this more dynamic. But basically it will work this way: the pkey will toggle between `01` and `07`. The client cert will toggle between `03` and `09`. 

I'll try and figure out a way to both not make it fixed (need flexibility) but also not make the user specify the slots each time (cause its too easy to screw up just like i did here).